httpbakery/client.go

```diff
@@ -343,14 +344,47 @@ func (c *Client) setRequestBody(req *http.Request, body io.ReadSeeker) error {
 	if body == nil {
 		return nil
 	}
-	if req.Body == nil {
-		req.Body = ioutil.NopCloser(body)
-	} else {
-		_, err := body.Seek(0, 0)
-		if err != nil {
+	if req.Body != nil {
+		req.Body.Close()
+		if _, err := body.Seek(0, 0); err != nil {
 			return errgo.Notef(err, "cannot seek to start of request body")
 		}
 	}
+	// Always replace the body with a new readStopper so that
+	// the old request cannot interfere with the new request's reader.
+	req.Body = &readStopper{
```

Comment:
That paragraph from the NewRequest godoc is not relevant to us here, because we don't allow a Body to be passed in the request (we actually check and return an error if that's the case). It would be surprising if the body was closed, as the body passed to DoWithBody is not a Closer.

The NewRequest behaviour is an acknowledged bad design - it just cannot be changed due to backward-compatibility concerns.